src/mlpack/methods/ann/layer/multihead_attention_impl.hpp

```diff
@@ -0,0 +1,320 @@
+  arma::mat zQuery = boost::apply_visitor(outputParameterVisitor, queryLinear);
+  arma::mat zKey = boost::apply_visitor(outputParameterVisitor, keyLinear);
+  arma::mat zValue = boost::apply_visitor(outputParameterVisitor, valueLinear);
+  zQuery /= std::sqrt(headDim); // Scaling factor.
+
+  // Check if 2D mask is provided.
+  if (!attnMask2d.is_empty())
+  {
+    //! If 2D mask is provided, check for the validity of constructor arguments.
+    if (attnMask2d.n_rows != tgtSeqLen || attnMask2d.n_cols != srcSeqLen)
+      Log::Fatal("The size of the 2D `attn_mask` is not correct.");
+  }
+  // Check if 3D mask is provided instead.
+  else if (!attnMask3d.is_empty())
+  {
+    // So if 3D mask is provided, check for corresponding constructr arguments.
+    if (attnMask3d.n_rows != tgtSeqLen || attnMask3d.n_cols != srcSeqLen ||
+        attnMask3d.n_slices != batchSize * numHeads)
+      Log::Fatal("The size of the 2D `attn_mask` is not correct.");
+  }
+
+  // Check if bias has to be appended to key and value matrices.
+  if (appendBias)
+  {
+    // Since we will be appending `colvec` or a slice to some of the parameters.
+    // New shape of these parameters will be mutated. So, effectively value of
+    // `srcSeqLen` is incremented by 1.
+    srcSeqLen += 1; 
+    // Extend the bias to `batchSize` and then append it to `zKey` and `zValue`.
+    zKey = arma::join_cols(zKey, arma::repmat(kBias, 1, batchSize));
+    // Now the shape of `zKey` is (srcSeqLen * embedDim, batchSize).
+    zValue = arma::join_cols(zValue, arma::repmat(vBias, 1, batchSize));
+    // Now the shape of `zValue` is (srcSeqLen * embedDim, batchSize).
+    // Check if 2D attention mask is specified.
+    if (!attnMask2d.is_empty())
+    {
+      // If so, append a column of zeros at the end.
+      attnMask2d = arma::join_rows(attnMask2d, 
+          arma::zeros(attnMask2d.n_rows, 1));
+      // Now the shape of `attnMask2d` is (tgtSeqLen, srcSeqLen).
+    }
+    // Check if 3D attention mask is specified.
+    else if (!attnMask3d.is_empty())
+    {
+      // If so, append a slice of zeros at the back.
+      attnMask3d = arma::join_rows(attnMask3d,
+          arma::zeros(attnMask3d.n_rows, attnMask3d.n_cols));
+      // Now the shape of `attnMask3d` is 
+      // (tgtSeqLen, srcSeqLen, numHeads * batchSize).
+    }
+  }
+
+  // Reshape the activations to 3D tensors without copying.
+  arma::cube q(zQuery.memptr(), tarSeqLen, headDim, numHeads * batchSize, 0);
+  arma::cube k(zKey.memptr(), srcSeqLen, headDim, numHeads * batchSize, 0);
+  arma::cube v(zValue.memptr(), srcSeqLen, headDim, numHeads * batchSize, 0);
+
+  // Create a new 3D tensor to store the results of operations.
+  arma::cube attnOutputWeight(tarSeqLen, srcSeqLen, numHeads * batchSize);
+  for (size_t i = 0; i < numHeads; ++i)
+  {
+    // Multiply each slice of `q` and `k`. (Is there any better way to do this?)
+    attnOutputWeight.slice(i) = q.slice(i) * k.slice(i).t();
+  }
```

Comment:
@lozhnikov: Is there some way other than using for loop to multiply two `cubes`.